tokio-async-await/src/async_await/compat/backward.rs

```diff
@@ -0,0 +1,112 @@
+use futures::{
+    Future as Future01,
+    Poll as Poll01,
+};
+use futures_core::{Future as Future03};
+
+use std::boxed::PinBox;
+use std::future::FutureObj;
+use std::ptr::NonNull;
+use std::task::{
+    Context,
+    Spawn,
+    UnsafeWake,
+    LocalWaker,
+    Poll as Poll03,
+    Waker,
+    SpawnObjError,
+};
+
+/// Convert an 0.3 `Future` to an 0.1 `Future`.
+#[derive(Debug)]
+pub struct Compat<T>(PinBox<T>);
+
+impl<T> Compat<T> {
+    pub fn new(data: T) -> Compat<T> {
+        Compat(PinBox::new(data))
+    }
+}
+
+/// Convert a valuee into one that can be used with `await!`.
+pub trait IntoAwaitable {
+    type Awaitable;
+
+    fn into_awaitable(self) -> Self::Awaitable;
+}
+
+impl<T> IntoAwaitable for T
+where T: Future03,
+{
+    type Awaitable = Self;
+
+    fn into_awaitable(self) -> Self {
+        self
+    }
+}
+
+impl<T, Item, Error> Future01 for Compat<T>
+where T: Future03<Output = Result<Item, Error>>,
+{
+    type Item = Item;
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll01<Item, Error> {
+        use futures::Async::*;
+
+        let local_waker = noop_local_waker();
+        let mut executor = NoopExecutor;
+
+        let mut cx = Context::new(&local_waker, &mut executor);
+
+        let res = self.0.as_pin_mut().poll(&mut cx);
+
+        match res {
+            Poll03::Ready(Ok(val)) => Ok(Ready(val)),
+            Poll03::Ready(Err(err)) => Err(err),
```

Comment:
Nit, TIOLI: I think this could also be
```rust
match res {
    Poll03::Ready(r) => Ok(Ready(r?)),
    Poll03::Pending => Ok(NotReady),
}
```
although, I'm not sure if that really makes the code _significantly_ clearer.
```